### PR TITLE
First stab at deploying executable built locally with docker

### PR DESCRIPTION
**Pull request by abailly**
This is a first stab at using docker to build locally an executable to be run on AWS Lambda. 
Some notes:

* It uses my [system-extra](https://github.com/abailly/system-extra) package as a submodule to invoke docker
* The built package does not seem to work currently, here is the error I got when trying to invoke API:
```
$ curl -v -X POST -H "Content-Type: application/json" -d "{\"name\": \"cup\", \"shape\": \"round\", \"size\": 3}" "$API/things/cup"
*   Trying 54.192.185.94...
* Connected to lzrfju1akf.execute-api.us-east-1.amazonaws.com (54.192.185.94) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
* Server certificate: *.execute-api.us-east-1.amazonaws.com
* Server certificate: Symantec Class 3 Secure Server CA - G4
* Server certificate: VeriSign Class 3 Public Primary Certification Authority - G5
> POST /v1/things/cup HTTP/1.1
> Host: lzrfju1akf.execute-api.us-east-1.amazonaws.com
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 44
> 
* upload completely sent off: 44 out of 44 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 164
< Connection: keep-alive
< Date: Wed, 30 Nov 2016 09:54:03 GMT
< Access-Control-Allow-Headers: Content-Type,X-Amz-Date,Authorization,X-Api-Key
< Access-Control-Allow-Methods: DELETE,GET,HEAD,POST,PUT,OPTIONS,TRACE
< Access-Control-Allow-Origin: *
< x-amzn-RequestId: ed110adb-b6e2-11e6-8962-8bd0479d2e61
< X-Amzn-Trace-Id: Root=1-583ea1bb-d4590527a8a7716e6f2dc7fd
< X-Cache: Miss from cloudfront
< Via: 1.1 a98053a65d016ea1580cbb8422815b10.cloudfront.net (CloudFront)
< X-Amz-Cf-Id: f7J6Hun6w6CmMrF7tZgphjbLDtnDapNWFaXjPzOQ192Jmbw7ELcy4g==
< 
{"errorMessage":"Cannot find module '/var/task/index'","errorType":"Error","stackTrace":["Module.require (module.js:353:17)","require (internal/module.js:12:17)"]}
```

### CODE REVIEW

**Comment by ababkin** [comment]
awesome, looking into it. '/var/task/index' looks strange, googled around and looks like people get error messages with '/var/task/index.js' and not '/var/task/index'. This tells me that perhaps the .js ext is somehow missed somewhere.

**Comment by abailly** [comment]
Will investigate more tomorrow. It worked with my wrapper so it must be
something in the  package or the lambda config. The package is correctly
uploaded to S3, maybe it is not uploaded to lambda ?

**Comment by ababkin** [comment]
cloned your fork and getting the build error below. Investigating

```
--  While building package qmuli-0.0.1 using:
      /Users/ababkin/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.24.0.0-ghc-8.0.1 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.24.0.0 build lib:qmuli exe:apigw-lambda-dynamodb exe:apigw-lambda-s3 exe:simple-s3-copy --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/ababkin/work/abailly/qmuli/.stack-work/logs/qmuli-0.0.1.log

    Configuring qmuli-0.0.1...
    Preprocessing library qmuli-0.0.1...
    [ 1 of 29] Compiling Qi.Deploy.Build  ( src/Qi/Deploy/Build.hs, .stack-work/dist/x86_64-osx/Cabal-1.24.0.0/build/Qi/Deploy/Build.o )

    /Users/ababkin/work/abailly/qmuli/src/Qi/Deploy/Build.hs:9:44: error:
        Not in scope: type constructor or class ‘BuildTarget’
```

**Comment by ababkin** [comment]
good news: i managed to fix it, the lambda works as expected now

two things I did:
1. give permissions to execute the binary (chmod +x apigw-lambda-dynamodb) before zipping the package and uploading it to lambda
2. got rid of the double naming and just used the hardcoded string name (as I had before) for both appName and buildName. I don't know if the double name is really necessary, unless I am missing something in your original intent, IMO it looks cleaner without the extra two (potentially redundant) command line arguments and works perfectly.

I'm so excited that this worked and one does not need to spin up another box just to build the binary.
Thank you for this!

**Comment by abailly** [comment]
Great. So the appName must be the same as the executable name? That totally makes sense actually... I will modify the code in that direction. The intent is that the target you want to build might not be deployed with this name, if only because of conflicting names in AWS. It actually would be much better IMHO to generate a unique name by default which would ensure people can actually use the deployment confidently.

There is one thing that slightly annoys me with the current setup: The Lambda program is used both for deployment (as a `ConfigProgram` ?) and for actually running the code when invoked. Would it not make sense to separate qmuli in 2 libraries: One for managing the lambdas, the other for running them.

**Comment by abailly** [comment]
Apologies for the submodules mess: Yes, I forgot to update the subs configuration after pushing. Going to fast...

**Comment by abailly** [comment]
OK, so I fixed the permissions which and managed to get program executed, but ran into the following error:
```
2016-12-01T10:25:08.641Z 6ec43bac-b7b0-11e6-9512-5724ccc21ef6 input: " {\"body-json\":{\"name\":\"cup\",\"shape\":\"round\",\"size\":3},\"params\":{\"path\":{\"thingId\":\"cup\"},\"querystring\":{},\"header\":{\"Accept\":\"*/*\",\"CloudFront-Forwarded-Proto\":\"https\",\"CloudFront-Is-Desktop-Viewer\":\"true\",\"CloudFront-Is-Mobile-Viewer\":\"false\",\"CloudFront-Is-SmartTV-Viewer\":\"false\",\"CloudFront-Is-Tablet-Viewer\":\"false\",\"CloudFront-Viewer-Country\":\"FR\",\"Content-Type\":\"application/json\",\"Host\":\"n12yrc0k0g.execute-api.us-east-1.amazonaws.com\",\"User-Agent\":\"curl/7.43.0\",\"Via\":\"1.1 ec6a8fa210fe0949bc0ec0f33dd9d942.cloudfront.net (CloudFront)\",\"X-Amz-Cf-Id\":\"cPFMT6fptDpBT5I93k9UKDwM5JZqNyLXUEFBxwDodmovQB_FLiGhhw==\",\"X-Forwarded-For\":\"176.187.69.70, 54.240.150.61\",\"X-Forwarded-Port\":\"443\",\"X-Forwarded-Proto\":\"https\"}},\"stage-variables\":{},\"context\":{\"account-id\":\"\",\"api-id\":\"n12yrc0k0g\",\"api-key\":\"\",\"authorizer-principal-id\":\"\",\"caller\":\"\",\"cognito-authentication-provider\":\"\",\"cognito-authentication-type\":\"\",\"cognito-identity-id\":\"\",\"cognito-identity-pool-id\":\"\",\"http-method\":\"POST\",\"stage\":\"v1\",\"source-ip\":\"176.187.69.70\",\"user\":\"\",\"user-agent\":\"curl/7.43.0\",\"user-arn\":\"\",\"request-id\":\"6ec4150e-b7b0-11e6-b60b-3bb8906a6e05\",\"resource-id\":\"7srk1i\",\"resource-path\":\"/things/{thingId}\"}} "
2016-12-01T10:25:08.996Z 6ec43bac-b7b0-11e6-9512-5724ccc21ef6 stdout: "Invalid qmulus name: 'putThing', the name should only contain alphanumeric lower case characters or a hyphen
"
2016-12-01T10:25:08.996Z 6ec43bac-b7b0-11e6-9512-5724ccc21ef6 stderr: ""
```

I understand where it comes from: The `index.js` wrapper has wrong number of arguments because I have added the program and target names as arguments. I will modify `Qi.hs` to improve argument parsing so that those names can be left implicit or given as parameters to the command-line. I like optparse-applicative, hope you don't mind me adding it to the mix?

**Comment by ababkin** [comment]
I'll try to expand a bit on my original vision about the 'universal' binary, hopefully it clears up my intent:
I decided to generate only one binary for everything because 1. i thought it was simpler and 2. i don't even know how to generate multiple binaries with different behaviors from the same Main.hs, and the intent was to have a one unified codebase anyway.
So the idea was to have only one binary and "parametrize" its behavior, i.e with some options it will act as a deployment tool, yet with other options, it will act as one of lambda behaviors.
This approach worked fine before the dockerization, but of course, with addition of AWS arch specific docker build, we now have two binaries: one that will be run on a developer machine for deployments (and future static analyses) and such and the other that will actually power the Lambda, and i think they can be named the same (i know you mentioned potential naming conflict, but i cannot think of a situation where this would be detrimental. Can you elaborate on this concern and perhaps give an example?).
Later on, when it's time for AWS to execute a particular lambda, the index.js shim gets the lambda name from the context that AWS passes. That AWS lambda name contains the combination of appName and the name of the lambda - these two are used to determine what binary to call and with what parameters.
I think (and please correct me if i'm mistaken) that it is ok to build the same project on dev machine and for Lambda in docker and just deploy the Lambda version the way you do it now.
I honestly don't think the name parametrization is necessary - it just takes the name of the current executable as the appName - so there is no possibility of name confusion/mismatch.

Few words on `ConfigProgram` and `LambdaProgram` DSL ASTs, as I'm not too sure how understandable the code is in its current state.
Basically the deployment commands use  the `ConfigProgram` part to generate complete cloudformation configuration in the qmuli's internal data structures and then generate a CF template off of those structures using `stratosphere` package.
Once a particular lambda is executed (with appropriate parameters passed), the binary interprets that particular `LambdaProgram` in context of the `ConfigProgram` (so lambda behavior can be aware of the whole CF configuration) and turns it into an IO monadic chain to be executed.

Please let me know if this makes sense.

**Comment by abailly** [comment]
Hi Alex
I am on my phone so unable to give a thorough answer.

Yes, what your are saying is completely making sense and I think this is a
good way to separate things.

I did not delve into the details of the code as I am busy trying to make
your examples work with my docker build. Still not there on my branch but I
k is why :-)

Re. Names: I am not sure either what's The Right Thing™ to do. There is one
name which is used as a target to build executable, and one name used to
deploy to S3 and Lambda. My understanding is that the latter name must be
unique across AWS DC more precisely, the bucket name where the code is
uploaded for consumption by Lambda deployment must be unique. Using the
same name as the stack target might lead to frequent conflicts, esp. for
newcomers that will reuse some existing code base. For example, I were not
able to deploy one example initially after you reported your success as the
name 'aws-lambda-dynamodb' was in use (probably by you...).

It seems however that the name of the executable inside the lambda package
is not shared, nor is the name of the function itself (?).

How about having a single name (the target to build) but:

1/ upload to a unique directory in S3 -> no risk of conflicts
2/ package the binary under a standard name (eg. lambda.exe) -> simplifies
index.js wrapper

As a side note, do you know how to get the API ID without having to go the
console? Would be great to output that as an output of 'cf create' command.

**Comment by abailly** [comment]
Executable now runs properly but I get an error when trying to add a `thing` to the DB:

```
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 3090
< Connection: keep-alive
< Date: Thu, 01 Dec 2016 18:37:24 GMT
< Access-Control-Allow-Headers: Content-Type,X-Amz-Date,Authorization,X-Api-Key
< Access-Control-Allow-Methods: DELETE,GET,HEAD,POST,PUT,OPTIONS,TRACE
< Access-Control-Allow-Origin: *
< x-amzn-RequestId: 33655bdd-b7f5-11e6-8a72-6f0298c56119
< X-Amzn-Trace-Id: Root=1-58406de3-b675ab13e9827153c946ba65
< X-Cache: Miss from cloudfront
< Via: 1.1 33be852abb39a8e95242a0c764cdc483.cloudfront.net (CloudFront)
< X-Amz-Cf-Id: p0RXjJk1M9L5qguQQkVR7IKZAwkI6fcvMy5IVdJyJdNqN5I3yvfN6A==
< 
{"errorMessage":"[500] Error: Command failed: /bin/sh -c ./myapigw-lambda-dynamodb foo bar lbd getThing \"{\\\"body-json\\\":{},\\\"params\\\":{\\\"path\\\":{\\\"thingId\\\":\\\"cup\\\"},\\\"querystring\\\":{},\\\"header\\\":{\\\"Accept\\\":\\\"*/*\\\",\\\"CloudFront-Forwarded-Proto\\\":\\\"https\\\",\\\"CloudFront-Is-Desktop-Viewer\\\":\\\"true\\\",\\\"CloudFront-Is-Mobile-Viewer\\\":\\\"false\\\",\\\"CloudFront-Is-SmartTV-Viewer\\\":\\\"false\\\",\\\"CloudFront-Is-Tablet-Viewer\\\":\\\"false\\\",\\\"CloudFront-Viewer-Country\\\":\\\"FR\\\",\\\"Host\\\":\\\"vrsaw2akb0.execute-api.us-east-1.amazonaws.com\\\",\\\"User-Agent\\\":\\\"curl/7.43.0\\\",\\\"Via\\\":\\\"1.1 33be852abb39a8e95242a0c764cdc483.cloudfront.net (CloudFront)\\\",\\\"X-Amz-Cf-Id\\\":\\\"lrqTOE656c9zcBFmVER1GCrZxLfiN7M-3P6t1WvsPmSts5xFnADKvw==\\\",\\\"X-Forwarded-For\\\":\\\"109.10.173.122, 54.239.156.112\\\",\\\"X-Forwarded-Port\\\":\\\"443\\\",\\\"X-Forwarded-Proto\\\":\\\"https\\\"}},\\\"stage-variables\\\":{},\\\"context\\\":{\\\"account-id\\\":\\\"\\\",\\\"api-id\\\":\\\"vrsaw2akb0\\\",\\\"api-key\\\":\\\"\\\",\\\"authorizer-principal-id\\\":\\\"\\\",\\\"caller\\\":\\\"\\\",\\\"cognito-authentication-provider\\\":\\\"\\\",\\\"cognito-authentication-type\\\":\\\"\\\",\\\"cognito-identity-id\\\":\\\"\\\",\\\"cognito-identity-pool-id\\\":\\\"\\\",\\\"http-method\\\":\\\"GET\\\",\\\"stage\\\":\\\"v1\\\",\\\"source-ip\\\":\\\"109.10.173.122\\\",\\\"user\\\":\\\"\\\",\\\"user-agent\\\":\\\"curl/7.43.0\\\",\\\"user-arn\\\":\\\"\\\",\\\"request-id\\\":\\\"33655bdd-b7f5-11e6-8a72-6f0298c56119\\\",\\\"resource-id\\\":\\\"zmklba\\\",\\\"resource-path\\\":\\\"/things/{thingId}\\\"}}\"\nmyapigw-lambda-dynamodb: ServiceError (ServiceError' {_serviceAbbrev = Abbrev \"DynamoDB\", _serviceStatus = Status {statusCode = 400, statusMessage = \"Bad Request\"}, _serviceHeaders = [(\"Server\",\"Server\"),(\"Date\",\"Thu, 01 Dec 2016 18:37:24 GMT\"),(\"Content-Type\",\"application/x-amz-json-1.0\"),(\"Content-Length\",\"112\"),(\"Connection\",\"keep-alive\"),(\"x-amzn-RequestId\",\"79G0AK2ITTTR3HE77CJCVCQE47VV4KQNSO5AEMVJF66Q9ASUAAJG\"),(\"x-amz-crc32\",\"3737639027\")], _serviceCode = ErrorCode \"ResourceNotFound\", _serviceMessage = Just (ErrorMessage \"Requested resource not found\"), _serviceRequestId = Just (RequestId \"79G0AK2ITTTR3HE77CJCVCQE47VV4KQNSO5AEMVJF66Q9ASUAAJG\")})\n\", stderr: \"myapigw-lambda-dynamodb: ServiceError (ServiceError' {_serviceAbbrev = Abbrev \"DynamoDB\", _serviceStatus = Status {statusCode = 400, statusMessage = \"Bad Request\"}, _serviceHeaders = [(\"Server\",\"Server\"),(\"Date\",\"Thu, 01 Dec 2016 18:37:24 GMT\"),(\"Content-Type\",\"application/x-amz-json-1.0\"),(\"Content-Length\",\"112\"),(\"Connection\",\"keep-alive\"),(\"x-amzn-RequestId\",\"79G0AK2ITTTR3HE77CJCVCQE47VV4KQNSO5AEMVJF66Q9ASUAAJG\"),(\"x-amz-crc32\",\"3737639027\")], _serviceCode = ErrorCode \"ResourceNotFound\", _serviceMessage = Just (ErrorMessage \"Requested resource not found\"), _serviceRequestId = Just (RequestId \"79G0AK2ITTTR3HE77CJCVCQE47VV4KQNSO5AEMVJF66Q9ASUAAJG\")})\n\""}
* Connection #0 to host vrsaw2akb0.execute-api.us-east-1.amazonaws.com left intact
```

**Comment by ababkin** [comment]
Hmm strange, looks like it's getting 400 when trying to reach dynamodb. Will look into it tonight

**Comment by ababkin** [comment]
Arnaud,

I wanted to reproduce your 400 error but before this i wanted to try and use a static build target name 'lambda' as you've suggested above. I got stuck at this error and cannot easily figure out what causes it or how to fix it. 
I have committed my changes to your PR into your fork, please let me know what you think and feel free to change or get rid of the commit.

Maybe the error would be more obvious to you:
```
Error parsing targets: Component CExe "lambda" does not exist in package qmuli
apigw-lambda-dynamodb: callProcess: docker "run" "--cidfile=.cidfile" "-v" "/Users/ababkin/work/abailly/qmuli:/build" "--volumes-from=e8053bfcc8d71022aaea6174389f4b792ceac754d1f6c64bb44b8ca9c0e9cb24" "-v" "/root/.stack" "-w" "/build" "ghc-centos:qmuli" "stack" "build" "--allow-different-user" "qmuli:exe:lambda" (exit 1): failed
```

**Comment by ababkin** [comment]
Also I plan to add the API URL resulting from creating the CF stack in the near future - definitely a much needed feature

**Comment by abailly** [comment]
The builder tries to build something called `lambda` but this does not exist in the package.yaml file. The lambda should actually be harwired in the `copyFile` line in `Build.hs` and then you can use a single  `appName` as both the deploy name and the target name.

**Comment by abailly** [comment]
What do you think about the issue I mentionned on names conflict in AWS? I am not enough knowledgeable in the platform to understand if this conflict appears outside of S3.

**Comment by abailly** [comment]
Not having time now but I can fix the issue tomorrow.

**Comment by ababkin** [comment]
Sure, no rush
I definitely agree that it is an issue with global S3 bucket names having to be unique. 
It seems to me that having users specify a unique appName as a CLI argument is a good interim solution. And the deploy command should return an error if the appName is already "taken". 
I also prefix all of the resources' names with the appName, in case resource names are reused for different qmuli projects within one account.
Also, as i mentioned, I like the idea of using just hardcoded 'lambda' for the build name, since it will be the only binary in the Lambda zip package and is isolated from everything else.

**Comment by ababkin** [comment]
ok, got it to work on my end. Now one only needs to specify the unique app name as the first argument for the whole thing to work. Let me know what you think

**Comment by abailly** [comment]
Yeah, everything is working fine for me too! I completely overlooked the fact we were executing the program we needed to build! Well done :) Need to write some doc now...

**Comment by ababkin** [comment]
Yes, docs will be helpful for sure.
Btw, what would do you think it makes sense to add to the framework next?
There are definitely many things that would add a lot of value, for the next avenues to explore i was thinking:
1. Authentication - pretty important for securing the api endpoints. I was thinking to either use AWS Cognito or integrate with Auth0 (which i've used before)
2. Test suite - to start with I would add tests that assert correct CF template generation and that all the expected CF resources are added.
3. Smaller conveniences such as outputting API URL after stack creation

**Comment by abailly** [comment]
Le 3 déc. 2016 16:52, "Alex Babkin" <notifications@github.com> a écrit :

Yes, docs will be helpful for sure.
Btw, what would do you think it makes sense to add to the framework next?
There are definitely many things that would add a lot of value, for the
next avenues to explore i was thinking:

   1. Authentication - pretty important for securing the api endpoints. I
   was thinking to either use AWS Cognito or integrate with Auth0 (which i've
   used before)

Never used either one of them. If you have experience with Auth0 that would
speed up things


   1. Test suite - to start with I would add tests that assert correct CF
   template generation and that all the expected CF resources are added.

+1 I need to have a deeper look but I guess that if config and programs are
expressed in a free monad, it should not be difficult to write tests.

Potential writers of qmuli apps should understand how to write tests.


   1. Smaller conveniences such as outputting API URL after stack creation

+1
I think the next thing to have is a showcase application, something bigger
that would highlight the advantage (and possible shortcomings) of using
lambda and qmuli.

 Not sure what it could be though...



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/qmuli/qmuli/pull/5#issuecomment-264647001>, or mute the
thread

**Comment by ababkin** [comment]
1. I'm actually thinking to start with Cognito as it is a native AWS service so it may be easier to integrate (although it looks like i will need to create custom CF resource provider to use it with CF). It also seems to be cheaper than Auth0 once the app grows to many users.
2. They actually use `operational` monad, which is very similar to free but IMO easier to deal with. Yes, the idea was to 'defer' IO to as low abstraction level as possible so it is easier to test all layers above the lowest IO one.
3. Not sure if i mentioned but i have a minimal frontend to the apigw-lambda-dynamodb example here: https://github.com/ababkin/qmuli-example-frontend
You absolutely right about the showcase app - I'm working on a first serious qmuli app with an idea that developing it will provide the feedback that will drive the framework's features.
https://github.com/ababkin/followup

**Comment by abailly** [comment]
Cool stuff. As i said I really need to understand how to build lambdas with
qmuli which means I need to find a good case. Thinking about some stuff at
work on images transformations.

**Comment by ababkin** [comment]
Great! and of course, don't hesitate to let me know if you have any questions
Also please let me know when you want me to merge this PR. Can't wait to incorporate this docker build step into my workflow :)

**Comment by ababkin** [comment]
I'm going to merge this for now, we can always add the docs in the later PRs
